sql/src/main/java/io/crate/analyze/MultiSourceSelect.java

```diff
@@ -118,24 +91,64 @@ public QuerySpec querySpec() {
         return querySpec;
     }
 
-    private RelationSplitter initSources(Map<QualifiedName, AnalyzedRelation> sources) {
+    private HashMap<QualifiedName, Source> initializeSources(Map<QualifiedName, AnalyzedRelation> originalSources) {
+        HashMap<QualifiedName, Source> sources = new LinkedHashMap<>(originalSources.size());
+        for (Map.Entry<QualifiedName, AnalyzedRelation> entry : originalSources.entrySet()) {
+            Source source = new Source(entry.getValue());
+            sources.put(entry.getKey(), source);
+        }
+        return sources;
+    }
 
-        RelationSplitter splitter = RelationSplitter.process(querySpec, sources.values());
-        for (Map.Entry<QualifiedName, AnalyzedRelation> entry : sources.entrySet()) {
-            QuerySpec spec = splitter.getSpec(entry.getValue());
+    public void pushDownQuerySpecs() {
+        splitter.process();
+        for (Source source : sources.values()) {
+            QuerySpec spec = splitter.getSpec(source.relation());
             assert spec != null;
-            Source source = new Source(entry.getValue(), spec);
-            this.sources.put(entry.getKey(), source);
+            source.querySpec(spec);
         }
-        return splitter;
     }
 
     /**
      * Returns an orderBy containing only orderings which are not already pushed down to sources. The optional is only
-     * pressent if there are orderBySymbols left. The existing orderBy is returned if it
+     * present if there are orderBySymbols left. The existing orderBy is returned if it
      * there are no changes.
      */
     public Optional<OrderBy> remainingOrderBy() {
         return splitter.remainingOrderBy();
     }
+
+    public static class Source {
+        private final AnalyzedRelation relation;
+        private QuerySpec querySpec;
+
+        public Source(AnalyzedRelation relation) {
+            this(relation, null);
+        }
+
+        public Source(AnalyzedRelation relation, QuerySpec querySpec) {
```

Comment:
had to keep this constructor because it's on [ManyTableConsumer](https://github.com/crate/crate/blob/a/multi-source-merge/sql/src/main/java/io/crate/planner/consumer/ManyTableConsumer.java#L198)